test/ocsp_test.go

```diff
@@ -964,6 +964,264 @@ func TestOCSPReloadRotateTLSCertEnableMustStaple(t *testing.T) {
+	setOCSPStatus(t, addr, "configs/certs/ocsp/server-status-request-url-08-cert.pem", ocsp.Good)
+
+	// Store Dirs
+	storeDirA := createDir(t, "_ocspA")
+	defer removeDir(t, storeDirA)
+	storeDirB := createDir(t, "_ocspB")
+	defer removeDir(t, storeDirB)
+	storeDirC := createDir(t, "_ocspC")
+	defer removeDir(t, storeDirC)
+
+	// Seed server configuration
+	srvConfA := `
+		host: "127.0.0.1"
+		port: -1
+
+		tls {
+			cert_file: "configs/certs/ocsp/server-status-request-url-01-cert.pem"
+			key_file: "configs/certs/ocsp/server-status-request-url-01-key.pem"
+			ca_file: "configs/certs/ocsp/ca-cert.pem"
+			timeout: 5
+		}
+		store_dir: "%s"
+		cluster {
+			name: AB
+			host: "127.0.0.1"
+			advertise: localhost
+			port: -1
+
+			tls {
+				cert_file: "configs/certs/ocsp/server-status-request-url-02-cert.pem"
+				key_file: "configs/certs/ocsp/server-status-request-url-02-key.pem"
+				ca_file: "configs/certs/ocsp/ca-cert.pem"
+				timeout: 5
+			}
+		}
+	`
+	srvConfA = fmt.Sprintf(srvConfA, storeDirA)
+	sconfA := createConfFile(t, []byte(srvConfA))
+	defer removeFile(t, sconfA)
+	srvA, optsA := RunServerWithConfig(sconfA)
+	defer srvA.Shutdown()
+
+	// The rest
+	srvConfB := `
+		host: "127.0.0.1"
+		port: -1
+
+		tls {
+			cert_file: "configs/certs/ocsp/server-status-request-url-03-cert.pem"
+			key_file: "configs/certs/ocsp/server-status-request-url-03-key.pem"
+			ca_file: "configs/certs/ocsp/ca-cert.pem"
+			timeout: 5
+		}
+		store_dir: "%s"
+		cluster { 
+			name: AB
+			host: "127.0.0.1"
+			advertise: localhost
+			port: -1
+
+			routes: [ nats://localhost:%d ]
+
+			tls {
+				cert_file: "configs/certs/ocsp/server-status-request-url-04-cert.pem"
+				key_file: "configs/certs/ocsp/server-status-request-url-04-key.pem"
+				ca_file: "configs/certs/ocsp/ca-cert.pem"
+				timeout: 5
+			}
+		}
+	`
+	srvConfB = fmt.Sprintf(srvConfB, storeDirB, optsA.Cluster.Port)
+	conf := createConfFile(t, []byte(srvConfB))
+	defer removeFile(t, conf)
+	srvB, optsB := RunServerWithConfig(conf)
+	defer srvB.Shutdown()
+
+	// Client connects to server A.
+	cA, err := nats.Connect(fmt.Sprintf("tls://localhost:%d", optsA.Port),
+		nats.Secure(&tls.Config{
+			VerifyConnection: func(s tls.ConnectionState) error {
+				if s.OCSPResponse == nil {
+					return fmt.Errorf("missing OCSP Staple from server")
+				}
+				return nil
+			},
+		}),
+		nats.RootCAs(caCert),
+		nats.ErrorHandler(noOpErrHandler),
+	)
+	if err != nil {
+		t.Fatal(err)
+	}
+	checkClusterFormed(t, srvA, srvB)
+
+	// Revoke the seed server cluster certificate, following servers will not be able to verify connection.
+	setOCSPStatus(t, addr, "configs/certs/ocsp/server-status-request-url-02-cert.pem", ocsp.Revoked)
+
+	// Original set of servers still can communicate to each other, even though the cert has been revoked.
+	// NOTE: Should we unplug from the cluster in case our server is revoke and OCSP policy is always or must?
```

Comment:
In case a certificate has a staple that has been revoked at some point, currently it is not disconnecting itself from the other routes, should we make this the default maybe? /cc @philpennock

Reply:
This should be _consistent_ with what we do when a TLS certificate expires.  If a TLS certificate expires, do we drop the connections?  If yes, then an expired OCSP staple should drop too.  If no, then OCSP should probably not be more strict.

An architecture decision on "NATS: do TLS expiration dates apply only at connection establishment or provide an upper bound on when data can be passed under that certificate" should be made, and then enforced consistently.

I think we should probably use level-enforcement not edge-enforcement; I haven't checked but think we're probably edge-enforcement for TLS cert expiration now; so in that case I think commit the OCSP work as-is, consistent with the TLS enforcement, but separately have a new ADR to document how we think TLS certs should be enforced throughout the NATS ecosystem, including in clients across multiple languages, and then in separate work move to enforce that consistent.

Reply:
iirc in case a cert expires, established connections are not disconnected but new ones would refuse attempting to connect so this would be similar behavior to that, new cluster routes clients would refuse to connect to the server with the invalid staple.